npm_and_yarn/lib/dependabot/npm_and_yarn/version.rb

```diff
@@ -42,29 +55,58 @@ def self.semver_for(version)
         version
       end
 
-      sig { override.params(version: T.any(String, Gem::Version)).void }
+      sig do
+        override
+          .params(
+            version: T.any(
+              String,
+              Integer,
+              Float,
+              Gem::Version,
+              NilClass
+            )
+          )
+          .void
+      end
       def initialize(version)
         @version_string = T.let(version.to_s, String)
         version = version.gsub(/^v/, "") if version.is_a?(String)
 
         version, @build_info = version.to_s.split("+") if version.to_s.include?("+")
 
-        super
+        super(T.must(version))
+      end
+
+      sig do
+        override
+          .params(
+            version: T.any(
+              String,
+              Integer,
+              Float,
+              Gem::Version,
+              NilClass
+            )
+          )
+          .returns(Dependabot::NpmAndYarn::Version)
+      end
+      def self.new(version)
+        T.cast(super, Dependabot::NpmAndYarn::Version)
       end
 
       sig { returns(Integer) }
       def major
-        @major ||= T.let(segments[0] || 0, T.nilable(Integer))
+        @major ||= T.let(segments[0].to_i, T.nilable(Integer))
```

Comment:
Here and in `#minor`, `#patch`, I'm wondering why the `T.nilable`? Seems like these values will always be an integer.

Reply:
It's because of the use of `||=` syntax. It's shorthand for `@major || major = segments[0].to_i`, which implies that `@major` can be `nil`, so Sorbet forces us to make it `T.nilable(Integer)`.

It's accurate though, because `@major` is `nil` until we first call `major`. At which point we set `@major` and return it.

Reply:
Ok, thanks for the explanation. I think I'm just still getting the hang of sorbet. I was reading that as the type for the `segments[0].to_i` expression, which will always be an integer because `nil.to_i => 0` and `"xyz".to_i => 0`.